### ISSUE TITLE
threads callable does not work

### ISSUE THREAD

**Issue by rvalieris** [snakemake/snakemake#210]
**Snakemake version**
5.10.0

**Describe the bug**
on the docs it says I can use a callable on `threads:`, however it doesnt seems to work.

**Logs**
```
TypeError in line 20 of Snakefile:
int() argument must be a string, a bytes-like object or a number, not 'function'
  File "Snakefile", line 20, in <module>
```

**Minimal example**
```
rule a:
  threads: lambda wildcards: 1
  output: touch('a')
```

**Additional context**
the docs says:
```
Starting from version 3.7, threads can also be a callable that returns an int value.
The signature of the callable should be callable(wildcards[, input]).
```

**Comment by mwort** [comment]
I can confirm this. Upgraded from 5.6 to 5.10 and `threads: callable` gives me the described error.

**Comment by ManavalanG** [comment]
Doesn't work in v5.9.1 either :(

**Comment by whytestalyon** [comment]
**TL;DR**: broken in v5.9.0, fixed in v5.12.0

**Long Explanation**: In v5.9.1 the issue is stemming from `workflow.py`. There is a check indicating that a `callable` is an acceptable value however immediately following that there is a cast to an `int` causing the issue. 

```
if ruleinfo.threads:
    if (
        not isinstance(ruleinfo.threads, int)
        and not isinstance(ruleinfo.threads, float)
        and not callable(ruleinfo.threads)
    ):
        raise RuleException(
            "Threads value has to be an integer, float, or a callable.",
            rule=rule,
        )
    rule.resources["_cores"] = int(ruleinfo.threads)
```

Closer inspection shows that this was introduced by the merging of this [PR](https://github.com/snakemake/snakemake/pull/165) between the release of v5.8.2 and v5.9.0 .

Then it appears to have been resolved in the commit `9f2b27cd Fix thread functions` right before the release of v5.12.0 .

**Comment by whytestalyon** [comment]
@johanneskoester I apologize if this isn't the right way to flag this issue as potentially closable. I though if my comment above is correct then this is a ticket you can close 😁  and if not then more info on the issue is always appreciated.

**Comment by rvalieris** [comment]
@whytestalyon thanks for the heads up, I can confirm it works for me with version 5.22.1, so I will go ahead and close this.
